### PR TITLE
Implement PullUp Empty Results optimizer

### PR DESCRIPTION
**Pull request by Tmonster**
Currently this logic is sprinkled within filter pull up and filter push down. The problem is it's not at every operator, and adding empty results across multiple files within the filter pullup/pushdown operator can make the code hard to maintain and debug. 

This will unify the code and fix some fuzzer issues that are the cause of long execution times on the LHS of a join while a child on the RHS is empty . It can also help make other queries faster in a few cases.

One con is that some `ConversionErrors` will be hidden.

### CODE REVIEW

**Comment by Tmonster** [comment]
Main needs to be merged into feature, then the commit history should be cleaned up a little bit

**Comment by hannes** [comment]
Can you fix the conflict please?

File: src/include/duckdb/optimizer/empty_result_pullup.hpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/optimizer/column_binding_replacer.hpp"
+
+namespace duckdb {
+
+//! The Deliminator optimizer traverses the logical operator tree and removes any redundant DelimGets/DelimJoins
```
Comment (lnkuiper) [inline src/include/duckdb/optimizer/empty_result_pullup.hpp:15]: Could you update this comment to describe this operator?

File: src/include/duckdb/optimizer/empty_result_pullup.hpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+class EmptyResultPullup : LogicalOperatorVisitor {
+public:
+	EmptyResultPullup() {
+	}
+
+	unique_ptr<LogicalOperator> Optimize(unique_ptr<LogicalOperator> op);
+	unique_ptr<LogicalOperator> PullUpEmptyJoinChildren(unique_ptr<LogicalOperator> op);
```
Comment (lnkuiper) [inline src/include/duckdb/optimizer/empty_result_pullup.hpp:22]: Could you make `PullUpEmptyJoinChildren` private?

**Review by lnkuiper — CHANGES_REQUESTED** [review]
Thanks for the PR! This looks good, I just have some minor comments:

**Review by lnkuiper — APPROVED** [review]
Thanks for the changes, LGTM!

**Comment by Mytherin** [comment]
Thanks! Looks great

**Pull request merged**